plugins/inputs/prometheus/prometheus.go

```diff
@@ -441,6 +446,10 @@ func (p *Prometheus) gatherURL(u URLAndAddress, acc telegraf.Accumulator) error
 	if err != nil {
 		return fmt.Errorf("error reading body: %w", err)
 	}
+	if p.ContentLengthLimit != 0 && len(body) > p.ContentLengthLimit {
+		p.Log.Debugf("skipping %s: content length (%d) exceeded maximum body size (%d)", u.URL, resp.ContentLength, p.ContentLengthLimit)
+		return nil
+	}
```

Comment:
Isn't it better to check the size before reading the whole body?

You can use `resp.ContentLength` for that.

Reply:
Thank you for the reviews!

If exporters return responses using chunked encoding. The responses don't have a "Content-Length" header.
c.f. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#directives

In this case, we need to read the entire body to know the actual content length.

Reply:
Do not confuse this with the `Content-Length` HTTP header, see https://pkg.go.dev/net/http#Response, it says how many bytes may be read from `resp.Body`.

Reply:
Understood! Thank you for pointing that out. I will try using resp.ContentLength.

Reply:
Hi, @Hipska 
I'm sorry for not getting back to you sooner.

In my experiment, it seems that `resp.ContentLength` does not reflect the actual body length when an exporter responds with chunked encoding.

- Add debug messages to plugins/inputs/prometheus/prometheus.go as follows:

```go
 	bodyLength := len(body)
 	p.Log.Debugf("length of actual body: (%d)", bodyLength)
	p.Log.Debugf("value of \"resp.ContentLength\": (%d)", resp.ContentLength)
```

- Build and run the code with settings that scrape metrics from an exporter returning responses with chunked encoding. The output is as below:

```
2024-01-19T09:35:30Z D! [inputs.prometheus] length of actual body: (89)
2024-01-19T09:35:30Z D! [inputs.prometheus] value of "resp.ContentLength": (-1)
```

The value of `resp.ContentLength` is `-1`, which indicates the length is unknown.
Based on these results, I will proceed with implementing the actual response length as initially planned.

※ For your reference, I will note the code of the exporter server used in the experiment.

<details><summary>metrics-server.go</summary><div>

```go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
		flusher, _ := w.(http.Flusher)

		w.Header().Set("Transfer-Encoding", "chunked")

		// return sample metrics devided into three chunks
		fmt.Fprintln(w, "# HELP sample_counter_metrics")
		flusher.Flush()
		fmt.Fprintln(w, "# TYPE sample_counter_metrics")
		flusher.Flush()
		fmt.Fprintln(w, "sample_counter_metrics 100.0")
		flusher.Flush()
	})

	fmt.Println("Server started at http://localhost:8080")
	http.ListenAndServe(":8080", nil)
}

```
</div></details>

Reply:
Sure, go ahead then. The documentation of the field is a bit misleading then..